docs: fix incorrect on_task_start / on_task_complete signatures for multi-agent hooks (PR #1740)

## Context

[MervinPraison/PraisonAI PR #1740](https://github.com/MervinPraison/PraisonAI/pull/1740) was merged on 2026-05-30. It fixes the callback signature mismatch in `PraisonAIAgents.start()` / `AgentTeam.start()` verbose mode (the bug from issue #1730 — "[ERROR] Error on startup when using multi-agent").

After this fix, the **canonical signatures** of the multi-agent orchestration hooks are now confirmed in the SDK source as:

- `on_task_start(task, task_id)` — **2 positional arguments**
- `on_task_complete(task, task_output)` — **2 positional arguments**

These are the hooks attached at the `PraisonAIAgents` / `AgentTeam` level via the `hooks=MultiAgentHooksConfig(...)` parameter. Several pages in PraisonAIDocs currently document **incorrect signatures**, which would cause exactly the runtime `TypeError` the upstream PR just fixed.

> Note: this is distinct from the `Task`-level `on_task_complete=` parameter (set on the `Task(...)` constructor), which still takes a single `TaskOutput` argument. Those examples in the docs are correct and should NOT be changed.

---

## SDK Ground Truth (verified)

From `praisonai-package/src/praisonai-agents/praisonaiagents/agents/agents.py`:

```python
# Line 1310 — AgentTeam invokes the start hook with 2 args:
self.on_task_start(task, task_id)

# Line 1401 — AgentTeam invokes the complete hook with 2 args:
self.on_task_complete(task, task_output)

# Lines 1612–1631 — composed verbose-mode callbacks (the fix in PR #1740):
def composed_on_task_start(task, task_id):
    ...
    if original_on_task_start:
        original_on_task_start(task, task_id)

def composed_on_task_complete(task, task_output):
    ...
    if original_on_task_complete:
        original_on_task_complete(task, task_output)
```

From `praisonai-package/src/praisonai-agents/praisonaiagents/config/feature_configs.py` (around line 907):

```python
@dataclass
class MultiAgentHooksConfig:
    """
    Consolidates: completion_checker, on_task_start, on_task_complete
    """
    on_task_start: Optional[Callable] = None
    on_task_complete: Optional[Callable] = None
    completion_checker: Optional[Callable] = None
```

From `praisonai-package/src/praisonai-agents/praisonaiagents/task/task.py` line 63 (for contrast — Task-level is different):

```python
on_task_complete: Optional[Union[Callable[[TaskOutput], Any], ...]] = None
# Task-level still takes 1 arg: (task_output)
```

---

## Type of Change Required

**Content update only.** No new docs page is required. The SDK ground truth was already partially documented; we just need to correct signatures and clarify the two-tier callback model (multi-agent level vs task level).

---

## Files to Update

### 1. `docs/concepts/hooks.mdx` — **Multi-Agent Hooks section** (lines ~197–218)

**Current (incorrect):**

```python
from praisonaiagents import AgentTeam, MultiAgentHooksConfig

def on_task_start(task):                       # ❌ missing task_id
    print(f"Starting: {task.name}")

def on_task_complete(task, result):            # ❌ wrong param name
    print(f"Completed: {task.name}")

team = AgentTeam(
    agents=[agent1, agent2],
    hooks=MultiAgentHooksConfig(
        on_task_start=on_task_start,
        on_task_complete=on_task_complete,
        completion_checker=my_checker,
    )
)
```

**Replace with (correct):**

```python
from praisonaiagents import AgentTeam, MultiAgentHooksConfig

def on_task_start(task, task_id):
    print(f"Starting task {task_id}: {task.name}")

def on_task_complete(task, task_output):
    print(f"Completed: {task.name}")
    print(f"Output: {task_output.raw}")

team = AgentTeam(
    agents=[agent1, agent2],
    hooks=MultiAgentHooksConfig(
        on_task_start=on_task_start,
        on_task_complete=on_task_complete,
        completion_checker=my_checker,
    )
)
```

> Per AGENTS.md "Folder Placement Rules": `docs/concepts/` is HUMAN-APPROVED ONLY. Flag this update for human review before merging — but the content fix itself is mandatory because the current example will raise `TypeError` at runtime.

---

### 2. `docs/features/callbacks.mdx` — **"Verbose Mode and Process Orchestration" section** (lines ~955–1074)

This section mixes two callback APIs in one example, which caused the original confusion in issue #1730. It needs to be split clearly.

**Currently incorrect snippet (lines ~983–1011):**

```python
def task_start_callback(task):                        # ❌ wrong if used at team level
    print(f"Starting task: {task.description}")

def task_complete_callback(task_output):              # ❌ wrong if used at team level
    print(f"Completed task: {task_output.description}")
    print(f"Result: {task_output.raw[:100]}...")

task1 = Task(
    description="First task in workflow",
    agent=agent,
    on_task_start=task_start_callback,                # ❌ Task() has no on_task_start
    on_task_complete=task_complete_callback           # ❌ wrong arity for verbose composition
)
```

**Replace with two clearly-separated examples:**

```python
# ── Multi-agent verbose mode: callbacks composed at PraisonAIAgents level ──
from praisonaiagents import Agent, Task, PraisonAIAgents, MultiAgentHooksConfig

def on_task_start(task, task_id):
    print(f"[{task_id}] Starting: {task.description}")

def on_task_complete(task, task_output):
    print(f"Completed: {task.description}")
    print(f"Output: {task_output.raw[:100]}...")

agent = Agent(name="Workflow Agent", instructions="Execute tasks in workflow")

task1 = Task(description="First task", agent=agent)
task2 = Task(description="Second task", agent=agent, context=[task1])

agents = PraisonAIAgents(
    agents=[agent],
    tasks=[task1, task2],
    process="workflow",
    hooks=MultiAgentHooksConfig(
        on_task_start=on_task_start,
        on_task_complete=on_task_complete,
    ),
    verbose=True,  # composes user callbacks with verbose UI callbacks
)
agents.start()
```

And keep / clarify the Task-level example for the per-task `on_task_complete` (which keeps its `(task_output)` signature):

```python
# ── Task-level on_task_complete: still receives a single TaskOutput ──
def on_complete(task_output):
    print(f"Task done: {task_output.raw}")

task = Task(
    description="Important task",
    agent=agent,
    on_task_complete=on_complete,  # Task-level: 1 arg (task_output)
)
```

Also update the "Callback Composition" subsection (lines ~1024–1045) to make it explicit that:

- The composition happens at the `PraisonAIAgents` / `AgentTeam` level, not at the `Task` level
- User hooks supplied via `MultiAgentHooksConfig(on_task_start=..., on_task_complete=...)` MUST accept `(task, task_id)` and `(task, task_output)` respectively
- Errors in either user or verbose callback are logged at debug level and do not interrupt the task
- Mention that this was fixed in PR #1740 (callback signature mismatch when combining user hooks with `verbose=True`)

---

### 3. `docs/sdk/reference/praisonaiagents/classes/MultiAgentHooksConfig.mdx`

The auto-generated SDK reference currently says "No description available." for each property. Update the descriptions (and ideally the docstrings in `feature_configs.py` upstream, but for this issue, just the docs page):

```mdx
<ResponseField name="on_task_start" type="Optional[Callable[[Task, int], None]]">
  Called before each task starts. Signature: `(task, task_id)`. `task_id` is the integer assigned by AgentTeam to this task in the run.
</ResponseField>

<ResponseField name="on_task_complete" type="Optional[Callable[[Task, TaskOutput], None]]">
  Called after each task finishes. Signature: `(task, task_output)`. `task_output` is the `TaskOutput` produced by the agent for this task.
</ResponseField>

<ResponseField name="completion_checker" type="Optional[Callable]">
  Custom function to decide whether the team's run has finished. Overrides `AgentTeam.default_completion_checker`.
</ResponseField>
```

---

### 4. `docs/concepts/hooks-vs-callbacks.mdx` (lines ~155–230) — sanity check

The example at line 160 uses `on_task_complete=task_done` as a "callback" pattern. If `task_done` is shown anywhere with a different arity, align it to:

- Task-level: `def task_done(task_output): ...`
- Multi-agent level (via `MultiAgentHooksConfig`): `def task_done(task, task_output): ...`

Add a small callout (e.g., `<Note>`) that distinguishes the two arities so users hit the right one on first try.

---

### 5. (Optional but recommended) Add a troubleshooting blurb

In `docs/features/callbacks.mdx`, add a short subsection right after "Callback Composition":

> **Troubleshooting: `TypeError: ... takes N positional arguments but M were given`**
>
> If you see this when calling `PraisonAIAgents.start()` with `verbose=True` and a custom `on_task_start` / `on_task_complete` hook, your hook signature does not match the multi-agent contract. Use:
>
> - `on_task_start(task, task_id)`
> - `on_task_complete(task, task_output)`
>
> Per-`Task` callbacks (set on the `Task(...)` constructor) keep their original 1-argument signature: `on_task_complete(task_output)`.

This directly addresses the user-reported bug from #1730 (which the user described as "an error occurs immediately, and I'm not sure how to resolve it").

---

## Acceptance Criteria

- [ ] `docs/concepts/hooks.mdx` Multi-Agent Hooks example uses the correct `(task, task_id)` and `(task, task_output)` signatures
- [ ] `docs/features/callbacks.mdx` Verbose Mode section clearly separates **multi-agent-level** hooks (`MultiAgentHooksConfig`) from **task-level** callbacks (`Task(on_task_complete=...)`)
- [ ] `docs/sdk/reference/praisonaiagents/classes/MultiAgentHooksConfig.mdx` properties have real descriptions with full callable types
- [ ] `docs/concepts/hooks-vs-callbacks.mdx` does not contradict the corrected signatures
- [ ] (Optional) Troubleshooting note added explaining the `TypeError` symptom
- [ ] All code examples copy-paste-run without raising `TypeError`
- [ ] Mermaid diagrams retain the standard color scheme from AGENTS.md
- [ ] Frontmatter, Steps / AccordionGroup / CardGroup components preserved per AGENTS.md template

## References

- Upstream fix: [MervinPraison/PraisonAI#1740](https://github.com/MervinPraison/PraisonAI/pull/1740)
- Original user bug report: [MervinPraison/PraisonAI#1730](https://github.com/MervinPraison/PraisonAI/issues/1730)
- SDK source (callback invocation): `praisonai-package/src/praisonai-agents/praisonaiagents/agents/agents.py` lines 1307–1403, 1609–1646
- SDK source (config dataclass): `praisonai-package/src/praisonai-agents/praisonaiagents/config/feature_configs.py` around line 907
- SDK source (Task-level callback contrast): `praisonai-package/src/praisonai-agents/praisonaiagents/task/task.py` line 63

--- a/docs/concepts/hooks-vs-callbacks.mdx
+++ b/docs/concepts/hooks-vs-callbacks.mdx
@@ -157,9 +157,18 @@ def task_done(output):
 
 task = Task(
     description="Say hello",
-    on_task_complete=task_done  # Callback on completion
+    on_task_complete=task_done  # Task-level: 1 arg (task_output)
 )
 ```
+
+<Note>
+**Callback Signature Distinction:**
+
+- **Task-level callbacks** (set on `Task()` constructor): Use 1 argument - `on_task_complete(task_output)`
+- **Multi-agent level callbacks** (set via `MultiAgentHooksConfig`): Use 2 arguments - `on_task_start(task, task_id)` and `on_task_complete(task, task_output)`
+
+Choose the right arity to avoid `TypeError` during execution.
+</Note>
 
 ---
 

--- a/docs/features/callbacks.mdx
+++ b/docs/features/callbacks.mdx
@@ -981,66 +981,69 @@ graph LR
 Previously, verbose mode bypassed workflow and hierarchical processes. Now the execution goes through `run_all_tasks()` so `process="workflow"` and `process="hierarchical"` work correctly with verbose display:
 
 ```python
-from praisonaiagents import Agent, Task, PraisonAIAgents
+# ── Multi-agent verbose mode: callbacks composed at PraisonAIAgents level ──
+from praisonaiagents import Agent, Task, PraisonAIAgents, MultiAgentHooksConfig
 
-def task_start_callback(task):
-    print(f"Starting task: {task.description}")
+def on_task_start(task, task_id):
+    print(f"[{task_id}] Starting: {task.description}")
 
-def task_complete_callback(task_output):
-    print(f"Completed task: {task_output.description}")
-    print(f"Result: {task_output.raw[:100]}...")
+def on_task_complete(task, task_output):
+    print(f"Completed: {task.description}")
+    print(f"Output: {task_output.raw[:100]}...")
 
-agent = Agent(
-    name="Workflow Agent",
-    instructions="Execute tasks in workflow"
-)
+agent = Agent(name="Workflow Agent", instructions="Execute tasks in workflow")
 
-task1 = Task(
-    description="First task in workflow",
-    agent=agent,
-    on_task_start=task_start_callback,
-    on_task_complete=task_complete_callback
-)
+task1 = Task(description="First task", agent=agent)
+task2 = Task(description="Second task", agent=agent, context=[task1])
 
-task2 = Task(
-    description="Second task in workflow",
-    agent=agent, 
-    context=[task1],  # Depends on task1
-    on_task_start=task_start_callback,
-    on_task_complete=task_complete_callback
-)
-
-# Verbose mode now works with workflow processes
 agents = PraisonAIAgents(
     agents=[agent],
     tasks=[task1, task2],
     process="workflow",
-    verbose=True  # Uses callback composition for display
+    hooks=MultiAgentHooksConfig(
+        on_task_start=on_task_start,
+        on_task_complete=on_task_complete,
+    ),
+    verbose=True,  # composes user callbacks with verbose UI callbacks
 )
+agents.start()
+```
 
-result = agents.start()
+```python
+# ── Task-level on_task_complete: still receives a single TaskOutput ──
+def on_complete(task_output):
+    print(f"Task done: {task_output.raw}")
+
+task = Task(
+    description="Important task",
+    agent=agent,
+    on_task_complete=on_complete,  # Task-level: 1 arg (task_output)
+)
 ```
 
 ### Callback Composition
 
-User-supplied `on_task_start` and `on_task_complete` callbacks are composed with verbose ones - both run:
+The composition happens at the `PraisonAIAgents` / `AgentTeam` level when user hooks are supplied via `MultiAgentHooksConfig` with `verbose=True`:
 
-- **User callback** executes first
-- **Verbose callback** executes second  
-- Errors in either are logged at debug level and do not interrupt the task
+- **User hooks** execute first with correct signatures: `(task, task_id)` and `(task, task_output)`
+- **Verbose callbacks** execute second for UI display
+- Errors in either user or verbose callback are logged at debug level and do not interrupt the task
+- This callback signature mismatch was fixed in PR #1740 for proper verbose mode composition
 
 ```python
-def user_task_callback(task_output):
-    # User's custom logic
+def user_on_task_complete(task, task_output):
+    # User's custom logic - note the 2-argument signature for multi-agent level
     save_to_database(task_output)
-    send_notification(task_output.description)
+    send_notification(task.description)
 
 # Both user callback and verbose display will run
-task = Task(
-    description="Important task",
-    agent=agent,
-    on_task_complete=user_task_callback,  # User logic
-    verbose=True  # Also triggers verbose display
+agents = PraisonAIAgents(
+    agents=[agent],
+    tasks=[task],
+    hooks=MultiAgentHooksConfig(
+        on_task_complete=user_on_task_complete,  # Multi-agent level: 2 args
+    ),
+    verbose=True  # Composes user hooks with verbose display
 )
 ```
 
@@ -1072,6 +1075,19 @@ for config in configs:
     print(f"Running {config['description']} process...")
     agents.start()
 ```
+
+### Troubleshooting: TypeError with Hook Signatures
+
+<Note>
+**Error: `TypeError: ... takes N positional arguments but M were given`**
+
+If you see this when calling `PraisonAIAgents.start()` with `verbose=True` and a custom `on_task_start` / `on_task_complete` hook, your hook signature does not match the multi-agent contract. Use:
+
+- `on_task_start(task, task_id)`
+- `on_task_complete(task, task_output)`
+
+Per-`Task` callbacks (set on the `Task(...)` constructor) keep their original 1-argument signature: `on_task_complete(task_output)`.
+</Note>
 
 ## Best Practices
 

--- a/docs/sdk/reference/praisonaiagents/classes/MultiAgentHooksConfig.mdx
+++ b/docs/sdk/reference/praisonaiagents/classes/MultiAgentHooksConfig.mdx
@@ -28,16 +28,16 @@ graph LR
 
 ## Properties
 
-<ResponseField name="on_task_start" type="Optional">
-  No description available.
+<ResponseField name="on_task_start" type="Optional[Callable[[Task, int], None]]">
+  Called before each task starts. Signature: `(task, task_id)`. `task_id` is the integer assigned by AgentTeam to this task in the run.
 </ResponseField>
 
-<ResponseField name="on_task_complete" type="Optional">
-  No description available.
+<ResponseField name="on_task_complete" type="Optional[Callable[[Task, TaskOutput], None]]">
+  Called after each task finishes. Signature: `(task, task_output)`. `task_output` is the `TaskOutput` produced by the agent for this task.
 </ResponseField>
 
-<ResponseField name="completion_checker" type="Optional">
-  No description available.
+<ResponseField name="completion_checker" type="Optional[Callable]">
+  Custom function to decide whether the team's run has finished. Overrides `AgentTeam.default_completion_checker`.
 </ResponseField>
 
 <Accordion title="Internal & Generic Methods">

